pkg/querier/dispatcher.go

```diff
@@ -0,0 +1,380 @@
+		return mimirpb.QUERY_ERROR_TYPE_CANCELED
+	}
+
+	var queryTimeoutErr promql.ErrQueryTimeout
+	if errors.As(err, &queryTimeoutErr) {
+		return mimirpb.QUERY_ERROR_TYPE_TIMEOUT
+	}
+
+	var storageError promql.ErrStorage
+	if errors.As(err, &storageError) {
+		return mimirpb.QUERY_ERROR_TYPE_INTERNAL
+	}
+
+	return mimirpb.QUERY_ERROR_TYPE_EXECUTION
+}
+
+type httpResultStream struct {
+	w http.ResponseWriter
+}
+
+func (s *httpResultStream) Write(ctx context.Context, r *frontendv2pb.QueryResultStreamRequest) error {
+	b, err := r.Marshal()
+	if err != nil {
+		return err
+	}
+
+	if err := binary.Write(s.w, binary.LittleEndian, uint64(len(b))); err != nil {
+		return err
+	}
+
+	if _, err := s.w.Write(b); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type queryResponseWriter struct {
+	stream frontendv2pb.QueryResultStream
+	logger log.Logger
+}
+
+func (w *queryResponseWriter) Write(ctx context.Context, r querierpb.EvaluateQueryResponse) error {
+	return w.stream.Write(ctx, &frontendv2pb.QueryResultStreamRequest{
+		Data: &frontendv2pb.QueryResultStreamRequest_EvaluateQueryResponse{
+			EvaluateQueryResponse: &r,
+		},
+	})
+}
+
+func (w *queryResponseWriter) WriteError(ctx context.Context, typ mimirpb.QueryErrorType, msg string) {
+	writeErrorToStream(ctx, w.stream, typ, msg, w.logger)
+}
+
+type evaluationObserver struct {
+	w         *queryResponseWriter
+	nodeIndex int64 // FIXME: remove this once Evaluator supports multiple nodes and passes the node index to the methods below
+
+	originalExpression string
+}
+
+func (o *evaluationObserver) SeriesMetadataEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, series []types.SeriesMetadata) error {
+	defer types.SeriesMetadataSlicePool.Put(&series, evaluator.MemoryConsumptionTracker)
+
+	protoSeries := make([]querierpb.SeriesMetadata, 0, len(series))
+
+	for _, s := range series {
+		protoSeries = append(protoSeries, querierpb.SeriesMetadata{
+			Labels: mimirpb.FromLabelsToLabelAdapters(s.Labels),
+		})
+	}
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_SeriesMetadata{
+			SeriesMetadata: &querierpb.EvaluateQueryResponseSeriesMetadata{
+				NodeIndex: o.nodeIndex,
+				Series:    protoSeries,
+			},
+		},
+	})
+}
+
+func (o *evaluationObserver) InstantVectorSeriesDataEvaluated(ctx context.Context, evaluator *streamingpromql.Evaluator, seriesIndex int, seriesData types.InstantVectorSeriesData) error {
+	defer types.PutInstantVectorSeriesData(seriesData, evaluator.MemoryConsumptionTracker)
+
+	// TODO: batch up series to return, rather than sending each series one at a time?
+
+	return o.w.Write(ctx, querierpb.EvaluateQueryResponse{
+		Message: &querierpb.EvaluateQueryResponse_InstantVectorSeriesData{
+			InstantVectorSeriesData: &querierpb.EvaluateQueryResponseInstantVectorSeriesData{
+				NodeIndex:  o.nodeIndex,
+				Floats:     mimirpb.FromFPointsToSamples(seriesData.Floats),
```

Comment:
These conversion methods do unsafe casts. I think this use is safe because the message is immediately serialized and written before the series data is returned to the pool, right?

Reply:
Yep, that's right. I've added some comments to make this clear in 66dc838.